internal/config/incremental_sync.go

```diff
@@ -0,0 +1,247 @@
+package config
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/icinga/icinga-go-library/database"
+	"github.com/icinga/icinga-go-library/types"
+	"github.com/jmoiron/sqlx"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"time"
+)
+
+// IncrementalConfigurable specifies Getter methods required for types supporting incremental configuration loading.
+type IncrementalConfigurable[PK comparable] interface {
+	zapcore.ObjectMarshaler
+
+	// GetID returns the primary key value.
+	GetID() PK
+
+	// GetChangedAt returns the changed_at value.
+	GetChangedAt() types.UnixMilli
+
+	// IsDeleted returns if this entry was marked as deleted.
+	IsDeleted() bool
+}
+
+// IncrementalConfigurableValidatable defines a validation method for new and updated elements within incrementalFetch.
+type IncrementalConfigurableValidatable interface {
+	// IncrementalValidate validates and optionally return an error, resulting in aborting this entry.
+	IncrementalValidate() error
+}
+
+// incrementalFetch queries all recently changed elements of BaseT and stores them in changeConfigSetField.
+//
+// The RuntimeConfig.configChangeTimestamps map contains the last known timestamp for each BaseT table. Only those
+// elements where the changed_at SQL column is greater than the stored timestamp will be fetched and stored in the
+// temporary RuntimeConfig.configChange ConfigSet. Later on, incrementalApplyPending merges it into the main ConfigSet.
+func incrementalFetch[
+	BaseT any,
+	PK comparable,
+	T interface {
+		*BaseT
+		IncrementalConfigurable[PK]
+	},
+](ctx context.Context, tx *sqlx.Tx, r *RuntimeConfig, changeConfigSetField *map[PK]T) error {
```

Comment:
I really like this idea, but, unless I am missing some point, this does not work. Your suggestion would add another type parameter `TPtr` which is not a requirement of `T`, but the other way around. Thus, using `*T` as the type would not allow accessing the methods defined by the `IncrementalConfigurable[PK]` interface. The `TPtr` type parameter would be unused and marked as such by the compiler.

To play around with this and to demonstrate, I created a mock-up here: https://go.dev/play/p/3XVTLj3Mjgu

Please feel free to correct me if I have misunderstood your point.

Reply:
I've missed that problem but it can be worked around by casting to `TPtr` (no type assertion required):
```go
TPtr(t).GetPrimaryKey()
```

Now the question is whether this makes it ugly again.

Reply:
IMHO, this would make it even more complex to comprehend. Then I would keep it as it is.